testing info (#8279)
aws-amplify/docs: src/pages/[platform]/start/migrate-to-gen2/index.mdx

The most notable change between Amplify Gen 1 backends and Gen 2 backends is the use of infrastructure as code with TypeScript. Preparing your environment for migration involves converting the configuration of your live Gen 1 backend resources to their Gen 2 equivalent written using TypeScript.

<Callout info>

**Testers Only**

Prerequisites: 

- Node.js v18.17.1 or higher

> - [Local AWS CLI profile](/gen1/[platform]/start/getting-started/installation/) with `AdministratorAccess-Amplify` and `theAmplifyBackendDeployFullAccess` policies attached.

Reviewer: ```suggestion
- [Local AWS CLI profile](/gen1/[platform]/start/getting-started/installation/) with `AdministratorAccess-Amplify` and the `AmplifyBackendDeployFullAccess` policies attached.
```

</Callout>

<Callout info>

**Testers Only** 

For testing purposes, you will need to additionally install the tagged versions of the `@aws-amplify/cli` package.

```bash title="Terminal" showLineNumbers={false}
npm install @aws-amplify/cli@12.15.0-next-3.0
```

</Callout>

Then push your newly-created `migrate` environment to create a set of resources matching the configuration of your live environments:

<Callout info>

**Testers Only** 

> if you have API resources, you will need to enable the following feature flag in your `amplify/cli.json` file:

Reviewer: ```suggestion
If you have API resources, you will need to enable the following feature flag in your `amplify/cli.json` file:
```

super nit

npx @aws-amplify/migrate@0.1.0-next-3.0 to-gen-2 prepare
```

</Callout>

```bash title="Terminal" showLineNumbers={false}
npx @aws-amplify/migrate@latest to-gen-2 prepare
```

After this command completes successfully your `amplify/` directory is converted to an Amplify Gen 2 backend authored in TypeScript. In the next section we will walk through how to validate the generated code using [Amplify Gen 2's sandbox feature](/[platform]/deploy-and-host/sandbox-environments/setup/).

## Validating the preparation

The preparation command will add npm dependencies relevant to your Gen 2 backend in the root `package.json`. Before you begin with the validation, install dependencies:

```bash title="Terminal" showLineNumbers={false}
npm install
```

<Callout info>

**Testers Only** 

For testing purposes, you will need to additionally install the tagged versions of the `@aws-amplify/backend` and `@aws-amplify/backend-data` packages.

```bash title="Terminal" showLineNumbers={false}
npm install @aws-amplify/backend@0.0.0-test-20250307211308 @aws-amplify/backend-data@0.0.0-test-20250307211308
```

</Callout>

<Callout info>

**Testers Only**

**Auth**

> We will need to add the following manually to `backend.ts` file to set up the additional user pool client for your `Identity providers`. This is needed since Gen 2 only supports 1 user pool client out-of-the-box:

Reviewer 2: Should this be setting up the identity providers for the additional user pool client ? Also, I think we can remove the `addClient` code since we are generating it and just keep the one after `COPY BELOW SECTION`. We can mention where to add this in the file as part of this line.

npx @aws-amplify/migrate@0.1.0-next-3.0 to-gen-2 prepare
```

</Callout>

```bash title="Terminal" showLineNumbers={false}
npx @aws-amplify/migrate@latest to-gen-2 prepare
```

After this command completes successfully your `amplify/` directory is converted to an Amplify Gen 2 backend authored in TypeScript. In the next section we will walk through how to validate the generated code using [Amplify Gen 2's sandbox feature](/[platform]/deploy-and-host/sandbox-environments/setup/).

## Validating the preparation

The preparation command will add npm dependencies relevant to your Gen 2 backend in the root `package.json`. Before you begin with the validation, install dependencies:

```bash title="Terminal" showLineNumbers={false}
npm install
```

<Callout info>

**Testers Only** 

For testing purposes, you will need to additionally install the tagged versions of the `@aws-amplify/backend` and `@aws-amplify/backend-data` packages.

```bash title="Terminal" showLineNumbers={false}
npm install @aws-amplify/backend@0.0.0-test-20250307211308 @aws-amplify/backend-data@0.0.0-test-20250307211308
```

</Callout>

<Callout info>

**Testers Only**

**Auth**

> We will need to add the following manually to `backend.ts` file to set up the additional user pool client for your `Identity providers`. This is needed since Gen2 only supports 1 user pool client out-of-the box:

Reviewer: ```suggestion
We will need to add the following manually to `backend.ts` file to set up the additional user pool client for your `Identity providers`. This is needed since Gen 2 only supports 1 user pool client out-of-the-box:
```

        Comment: "Comment-xnibcobqircidcckd2zx2uzxgq-dev",
      },
    },
  ],
  schema: `
// Copy your Gen 1 schema here
  `,
});
```

The add the following to `backend.ts` file; you will need to disable the `sseSpecification` override for each table configured in Gen 1. If the table does have a `sseSpecification` enabled in Gen 1, you can skip this step.

```typescript
// do for each table
// default for SSE changed to from false -> true in gen 2
// Set the gen 2 table to sseEnabled: false so that the table definitions match
const todoTable = backend.data.resources.cfnResources.additionalCfnResources['Todo'];
todoTable.addOverride('Properties.sseSpecification', { sseEnabled: false });
```

</Callout>

{/* how to handle the thrown errors in the code */}
{/* porting custom resolvers, overridden resolvers */}
{/* porting custom resources themselves */}

<Callout info>

In order to provide a smoother experience with the migration tooling, we have added `throw new Error` messages to the generated code. This is to ensure that resources requiring manual changes are modified in your sandbox environment before the migration is executed.

> For example, if you have a NodeJS Lambda function In Gen 1, you will see the following error in your sandbox run:

Reviewer: ```suggestion
For example, if you have a Node.js Lambda function in Gen 1, you will see the following error in your sandbox run:
```

```bash title="Terminal" showLineNumbers={false} 
Source code for this function can be found in your Amplify Gen 1 Directory. See .amplify/migration/amplify/backend/function/app81ff331aPostConfirmation/src
```

you will need to copy the code from your Gen 1 backend and make any necessary changes to the code such as updating to TypeScript and updating the import statements. 

**Note:** You will need to install the Lambda dependencies in the project root `package.json` file.

</Callout>

```bash title="Terminal" showLineNumbers={false}
npx ampx sandbox
```

<Callout info>

> If you have not deployed a Amplify Gen 2 backend before, the process will redirect you the Amplify Console to bootstrap your region. Ensure you are logged with Admin credentials.
> If you do experience any errors, refer to the [troubleshooting section]([platform]/build-a-backend/troubleshooting/).

Reviewer: ```suggestion
If you have not deployed a Amplify Gen 2 backend before, the process will redirect you the Amplify Console to bootstrap your AWS account and region. Ensure you are signed in with Admin credentials.

If you experience any errors, refer to the [troubleshooting section]([platform]/build-a-backend/troubleshooting/).
```

for (const provider in providerSetupResult) {
    const providerSetupPropertyValue = providerSetupResult[provider];
    if (providerSetupPropertyValue.node && providerSetupPropertyValue.node.id.endsWith('IDP')) {
        userPoolClient.node.addDependency(providerSetupPropertyValue);
    }
}
```

**Data**

If you have GraphQL API, you would need to manually input your schema in `data/resource.ts` file.

```typescript
import { defineData } from "@aws-amplify/backend";
export const data = defineData({
  migratedAmplifyGen1DynamoDbTableMap: [
    {
      branchName: "sandbox",
      modelTableNameMap: {
        Post: "Post-xnibcobqircidcckd2zx2uzxgq-dev",
        Comment: "Comment-xnibcobqircidcckd2zx2uzxgq-dev",
      },
    },
  ],
  schema: `
// Copy your Gen 1 schema here
  `,
});
```

The add the following to `backend.ts` file; you will need to disable the `sseSpecification` override for each table configured in Gen 1. If the table does have a `sseSpecification` enabled in Gen 1, you can skip this step.

```typescript
// do for each table
// default for SSE changed to from false -> true in gen 2
// Set the gen 2 table to sseEnabled: false so that the table definitions match

> const todoTable = backend.data.resources.cfnResources.additionalCfnResources['Todo'];

Reviewer 3: This shouldn't be necessary anymore with the latest gen 2 snapshot release.

  --from <amplify-gen-1-stack-name> \
  --to <amplify-gen-2-stack-name>
```

</Callout>

```bash title="Terminal" showLineNumbers={false}
npx @aws-amplify/migrate@latest to-gen-2 execute \
  --from <amplify-gen-1-stack-name> \
  --to <amplify-gen-2-stack-name>
```

## Reverting the migration

<Callout info>

**Testers Only**

```bash title="Terminal" showLineNumbers={false}
npx @aws-amplify/migrate@0.1.0-next-3.0 to-gen-2 revert \
  --from <amplify-gen-2-stack-name> \
  --to <amplify-gen-1-stack-name>
```

</Callout>

```bash title="Terminal" showLineNumbers={false}
npx @aws-amplify/migrate@latest to-gen-2 revert \
  --from <amplify-gen-2-stack-name> \
  --to <amplify-gen-1-stack-name>
```

Then copy the `amplify/` directory from the `.amplify/migration` directory back into your project.

## Post-migration tasks

{/* a note about s3 bucket names */}
{/* ... anything else */}
{/* set up CI/CD */}
{/* deploy using backend-cli manually (for those not using Amplify to deploy today) */}

<Callout info>

**Testers Only**

Delete Cognito User pool custom domain from `amplify/backend.ts` as the domain already exists (This will not delete the domain from Cognito User pool):

```typescript

> (backend.auth.stack.node.children.find(child => child.node.id === 'amplifyAuth') as any).domainPrefix;

Reviewer 2: There is a `delete` keyword missing at the beginning of this line.
